Refactoring to expose method to get context queries with exclusions

justinrsweeney opened the pull request.

magibney commented on solr/core/src/java/org/apache/solr/search/facet/FacetProcessor.java:
```diff
@@ -190,16 +191,23 @@ private void handleDomainChanges() throws IOException {
+
+    // recompute the base domain
+    fcontext.base = fcontext.searcher.getDocSet(qlist);
+  }
+
+  protected List<Query> getContextQueries() {
     List<String> excludeTags = freq.domain.excludeTags;
 
     if (excludeTags == null || excludeTags.size() == 0) {
-      return;
+      return Collections.emptyList();
     }
 
     Set<Query> excludeSet = QueryUtils.getTaggedQueries(fcontext.req, excludeTags);
 
     if (excludeSet.isEmpty()) {
-      return;
+      return Collections.emptyList();
```
We can't shortcircuit here (this will probably fail existing solr tests, I think?). The reason stock solr can shortcircuit is that if there are no tag/exclusions, this facet just inherits parent `fcontext`'s `fcontext.base` DocSet. But with this PR in Solr we're recomputing `fcontext.base` every time (which we can/should avoid). It's also returning empty query list, which will end up calculating it wrong I think anyway (and empty query list is not what we want form the FunnelFacetProcessor side either).

It's a little tricky; we want to:
1. avoid recomputing `fcontext.base` if it's not necessary
2. ideally avoid building query list if the caller doesn't care about it (solr-internal for the the "no exclusion" case, and
3. always get query list (respecting exclusions) if the caller _does_ care about something other than computing `fcontext.base` (i.e., when called from FunnelFacetProcessor)

justinrsweeney replied:
That makes sense, I think I've resolved those issues now.

magibney replied:
Is there a way that we can avoid calling `QueryUtils.getTaggedQueries(...)` twice for the `handleFilterExclusions()` path?

justinrsweeney replied:
Just changed that to pass the excludeSet into getContextQueries(). If it is null, it will recompute so from FunnelFacetProcessor for example, we can just call getContextQueries(null).

magibney reviewed (changes requested):
(comments on lines of code)

claygoddard commented:
All looks good to me! I think there a couple minor unrelated formatting / code changes though?

magibney commented on solr/core/src/java/org/apache/solr/search/facet/FacetProcessor.java:
```diff
@@ -202,6 +202,22 @@ private void handleFilterExclusions() throws IOException {
+  }
+
+  private List<Query> getContextQueriesWithExclusions(Set<Query> excludeSet) {
+    if (excludeSet == null) {
+      excludeSet = QueryUtils.getTaggedQueries(fcontext.req, freq.domain.excludeTags);
+    }
+
```
Nice! but with fb7b94090f91f84eb63b19a064e0f934e98ed69d, we no longer need this check, right?

magibney commented:
> All looks good to me! I think there a couple minor unrelated formatting / code changes though?

Agreed ... I think the formatting changes look good (they basically bring the formatting of the FacetProcessor.java file in line with upstream). But the changes to the other two files are unrelated -- could those be separated into a different PR? They just look like compile fixes (presumably related to recently syncing with upstream branch_9x?).

magibney reviewed (approved):
LGTM!

justinrsweeney commented:
Reverted the other changes and moved those to a new PR here: https://github.com/cowpaths/fullstory-solr/pull/71, those were just additional leftover issues from changes we have on our fork impacted by the merge from upstream.
